homeassistant/components/integration/sensor.py

```diff
@@ -291,108 +385,113 @@ async def async_added_to_hass(self) -> None:
             self._attr_device_class = state.attributes.get(ATTR_DEVICE_CLASS)
             self._unit_of_measurement = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
 
-        @callback
-        def calc_integration(event: EventType[EventStateChangedData]) -> None:
-            """Handle the sensor state changes."""
-            old_state = event.data["old_state"]
-            new_state = event.data["new_state"]
+        if self._max_dt is not None:
+            self._schedule_max_dt_exceeded()
+            self.async_on_remove(self._cancel_max_dt_exceeded_callback)
 
-            if (
-                source_state := self.hass.states.get(self._sensor_source_id)
-            ) is None or source_state.state == STATE_UNAVAILABLE:
-                self._attr_available = False
-                self.async_write_ha_state()
-                return
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass,
+                [self._sensor_source_id],
+                self._integrate_on_state_change_and_max_dt
+                if self._max_dt is not None
+                else self._integrate_on_state_change,
+            )
+        )
 
-            self._attr_available = True
```

Comment:
availability should be set when the sensor is added and should follow the source, why remove ?

Reply:
This has been moved to line 446 and it is (like with the current implementation) set after it passes some validations checks on the source.

Reply:
it's in `_integrate_on_state_change` but this method is not called by `async_added_to_hass`

Reply:
Correct. But in the old situation it was part of the method
```python
@callback
def calc_integration(event: EventType[EventStateChangedData]) -> None:
```
which was defined in the scope of the method `async_added_to_hass`. But it was only executed as part of state changes.

Reply:
you are right, should probably fix this is another PR 😓

Reply:
Good to know! Availability was and is now set only if the source sends an update. We can address this method afterwards. There is also a legacy comment in the `async_added_to_hass`. I expect this PR makes it easier to address because of the extraction of the calculation logic.